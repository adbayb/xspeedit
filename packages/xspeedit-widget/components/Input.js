import React, { Component } from "react";
import PropTypes from "prop-types";

import Icon from "../components/Icon";

class Input extends Component {
	static propTypes = {
		label: PropTypes.string,
		onSubmit: PropTypes.func.isRequired,
		onValidate: PropTypes.func.isRequired
	};

	state = {
		input: ""
	};

	handleSubmit = e => {
		e.preventDefault();

		const { onValidate, onSubmit } = this.props;
		const { input } = this.state;

		if (!onValidate(input)) {
			return false;
		}

		onSubmit(input);
		this.setState(() => ({ input: "" }));

		return false;
	};

	handleChange = e => {
		const name = e.target.name;

		if (!name) {
			return false;
		}

		this.setState({ [name]: e.target.value });
		return true;
	};

	render() {
		const { label, onValidate } = this.props;
		const { input } = this.state;

		return (
			<form onSubmit={this.handleSubmit}>
				{label && <label htmlFor="boxesInput">{label}</label>}
				<input
					id="boxesInput"
					name="input"
					value={input}
					type="number"
					placeholder="Saisissez vos objets"
					onChange={this.handleChange}
				/>
				<button type="submit" disabled={!onValidate(input)}>
					<Icon className="button" name="enter" />
				</button>
				<style jsx>{`
					form {
						display: flex;
						flex: none;
						justify-content: space-between;
						background-color: #6c727f;
					}
					input {
						width: calc(100% - 50px);
						font-size: 0.9rem;
						margin: 0 0.5rem;
						-webkit-appearance: none;
						-moz-appearance: textfield;
						appearance: none;
						border: 0;
						background-color: transparent;
						color: white;
					}
					input::placeholder {
						color: currentColor;
					}
					button {
						padding: 0 20px;
						background: none;
						border: none;
						border-radius: 0;
					}
					button:not([disabled]) {
						background-color: #34b7b6;
					}
					label {
						display: flex;
						background-color: #565b68;
						margin: 0;
						padding: 1rem;
						align-items: center;
						justify-content: center;
					}
				`}</style>
			</form>
		);
	}
}

export default Input;
