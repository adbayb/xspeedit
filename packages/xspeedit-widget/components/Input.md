**Preview:**

```js
<Input
	label="My input"
	onSubmit={value => console.log("Submitted value = ", value)}
	onValidate={value => Boolean(value)}
/>;
```

**Example:**

```jsx static
import React from "react";

const Example = () => {
	const handleSubmit = (value) => {
		console.log("Submitted value = ", value);
	};

	const handleValidate = (value) => {
		return Boolean(value);
	};

	return (
		<Input label="My input" onSubmit={handleSubmit} onValidate={handleValidate} />
	);
};
```
