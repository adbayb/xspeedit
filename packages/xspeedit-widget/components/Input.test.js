import React from "react";
import Input from "./Input.js";

describe("Input", () => {
	it("should render correctly", () => {
		// GIVEN: notre composant:
		const tree = shallow(<Input label="My input" onSubmit={() => true} onValidate={() => true} />);
		// THEN: il doit s'afficher correctement en matchant son snapshot:
		expect(tree).toMatchSnapshot();
	});

	it("should handle submit correctly when submitted a valid value", () => {
		// GIVEN: notre composant:
		const tree = shallow(<Input label="My input" onSubmit={() => true} onValidate={() => true} />);
		const event = { preventDefault: jest.fn() };
		const instance = tree.instance();
		// WHEN: nous envoyons notre input:
		const ret = instance.handleSubmit(event);
		// THEN: le retour de fonction doit se faire correctement:
		expect(ret).toBe(false);
	});

	it("should handle submit correctly when submitted an invalid value", () => {
		// GIVEN: notre composant:
		const tree = shallow(<Input label="My input" onSubmit={() => true} onValidate={() => false} />);
		const event = { preventDefault: jest.fn() };
		const instance = tree.instance();
		// WHEN: nous envoyons notre input:
		const ret = instance.handleSubmit(event);
		// THEN: le retour de fonction doit se faire correctement:
		expect(ret).toBe(false);
	});

	it("should handle a valid change correctly", () => {
		// GIVEN: notre composant:
		const tree = shallow(<Input label="My input" onSubmit={() => true} onValidate={() => true} />);
		const value = "value";
		const event = { target: { name: "name", value } };
		const instance = tree.instance();
		// WHEN: nous envoyons notre input:
		const ret = instance.handleChange(event);
		// THEN: le state doit-être mis à jour correctement:
		expect(instance.state.name).toBe(value);
		expect(ret).toBe(true);
	});

	it("should handle an invalid change correctly", () => {
		// GIVEN: notre composant:
		const tree = shallow(<Input label="My input" onSubmit={() => true} onValidate={() => true} />);
		const event = { target: {} };
		const instance = tree.instance();
		// WHEN: nous envoyons notre input:
		const ret = instance.handleChange(event);
		// THEN: le state doit-être mis à jour correctement:
		expect(instance.state.name).toBe(undefined);
		expect(ret).toBe(false);
	});
});
