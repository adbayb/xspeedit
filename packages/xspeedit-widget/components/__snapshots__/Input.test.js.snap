// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
<form
  className="jsx-2110388131"
  onSubmit={[Function]}
>
  <label
    className="jsx-2110388131"
    htmlFor="boxesInput"
  >
    My input
  </label>
  <input
    className="jsx-2110388131"
    id="boxesInput"
    name="input"
    onChange={[Function]}
    placeholder="Saisissez vos objets"
    type="number"
    value=""
  />
  <button
    className="jsx-2110388131"
    disabled={false}
    type="submit"
  >
    <Icon
      className="button"
      height={24}
      name="enter"
      width={24}
    />
  </button>
  <JSXStyle
    css="form.jsx-2110388131{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex:none;-ms-flex:none;flex:none;-webkit-box-pack:justify;-webkit-justify-content:space-between;-ms-flex-pack:justify;justify-content:space-between;background-color:#6c727f;}input.jsx-2110388131{width:calc(100% - 50px);font-size:0.9rem;margin:0 0.5rem;-webkit-appearance:none;-moz-appearance:textfield;-webkit-appearance:none;-moz-appearance:none;appearance:none;border:0;background-color:transparent;color:white;}input.jsx-2110388131::-webkit-input-placeholder{color:currentColor;}input.jsx-2110388131::-moz-placeholder{color:currentColor;}input.jsx-2110388131:-ms-input-placeholder{color:currentColor;}input.jsx-2110388131::placeholder{color:currentColor;}button.jsx-2110388131{padding:0 20px;background:none;border:none;border-radius:0;}button.jsx-2110388131:not([disabled]){background-color:#34b7b6;}label.jsx-2110388131{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;background-color:#565b68;margin:0;padding:1rem;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-pack:center;-webkit-justify-content:center;-ms-flex-pack:center;justify-content:center;}"
    styleId="2110388131"
  />
</form>
`;
